src/software/ai/evaluation/calc_best_shot.h

```diff
@@ -10,6 +10,372 @@
+                                                        angle_seg.getAngleBottom()));
+                return;
+            }
+        }
+
+        this->taken_angle_segments.emplace_back(angle_seg);
+    }
+
+    AngleSegment getBiggestViableAngleSegment() override
+    {
+        std::sort(this->taken_angle_segments.begin(), this->taken_angle_segments.end(),
+                  [](AngleSegment a, AngleSegment b) -> bool {
+                      Angle a_angle = a.getAngleTop();
+                      Angle b_angle = b.getAngleTop();
+
+                      if (a_angle < b_angle)
+                      {
+                          return false;
+                      }
+                      else if (a_angle > b_angle)
+                      {
+                          return true;
+                      }
+
+                      return false;
+                  });
+
+
+        AngleSegment biggest_viable_angle_seg =
+            AngleSegment(Angle::zero(), Angle::zero());
+        if (this->taken_angle_segments.empty())
+        {
+            biggest_viable_angle_seg = AngleSegment(this->angle_seg.getAngleTop(),
+                                                    this->angle_seg.getAngleBottom());
+            return biggest_viable_angle_seg;
+        }
+
+        AngleSegment first_taken_angle_seg = this->taken_angle_segments.front();
+        if (first_taken_angle_seg.getAngleTop() < this->angle_seg.getAngleTop())
+        {
+            biggest_viable_angle_seg = AngleSegment(this->angle_seg.getAngleTop(),
+                                                    first_taken_angle_seg.getAngleTop());
+        }
+
+        AngleSegment last_taken_angle_seg = this->taken_angle_segments.back();
+        if (last_taken_angle_seg.getAngleBottom() > this->angle_seg.getAngleBottom())
+        {
+            AngleSegment viable_angle_seg = AngleSegment(
+                last_taken_angle_seg.getAngleBottom(), this->angle_seg.getAngleBottom());
+            if (viable_angle_seg.getDelta() > biggest_viable_angle_seg.getDelta())
+            {
+                biggest_viable_angle_seg = viable_angle_seg;
+            }
+        }
+
+        for (auto i = this->taken_angle_segments.begin();
+             i < this->taken_angle_segments.end() - 1; i++)
+        {
+            AngleSegment taken_angle_seg      = i[0];
+            AngleSegment next_taken_angle_seg = i[1];
+            AngleSegment viable_angle_seg     = AngleSegment(
+                taken_angle_seg.getAngleBottom(), next_taken_angle_seg.getAngleTop());
+            if (viable_angle_seg.getDelta() > biggest_viable_angle_seg.getDelta())
+            {
+                biggest_viable_angle_seg = viable_angle_seg;
+            }
+        }
+
+        return biggest_viable_angle_seg;
+    }
+};
+
+/**
+ * Represents an AngleMap that is for shooting on the friendly goal. The angles are
+ * described as going from 0 -> 2 * pi
+ */
+class FriendlyAngleMap : public AngleMap
```

Comment:
There is a lot of duplication between `EnemyAngleMap` and `FriendlyAngleMap` only to handle a slightly different set of angles, do you think you could just use one?

Reply:
done